packages/csvviewer/src/model.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  DataModel
+} from '@phosphor/datagrid';
+
+
+/**
+ * A data model implementation for in-memory DSV data.
+ */
+export
+class DSVModel extends DataModel {
+/**
+ * Create a data model with static CSV data.
+ *
+ * @param options - The options for initializing the data model.
+ */
+constructor(options: DSVModel.IOptions) {
+  super();
+  this._data = options.data;
+  this._delimiter = options.delimiter;
+  this._computeOffsets();
+}
+
+/**
+ * Get the row count for a region in the data model.
+ *
+ * @param region - The row region of interest.
+ *
+ * @returns - The row count for the region.
+ */
+rowCount(region: DataModel.RowRegion): number {
+  if (region === 'body') {
+    return this._offsets.length;
+  }
+  return 1;  // TODO multiple column-header rows?
+}
+
+/**
+ * Get the column count for a region in the data model.
+ *
+ * @param region - The column region of interest.
+ *
+ * @returns - The column count for the region.
+ */
+columnCount(region: DataModel.ColumnRegion): number {
+  // Parse first row, return that number of columns
+    if (region === 'body') {
+      if (this._offsets.length > 1) {
+        return this._data.slice(0, this._offsets[1]).split(this._delimiter).length;
+      }
+    }
+    return 0;
+}
+
+/**
+ * Get the data value for a cell in the data model.
+ *
+ * @param region - The cell region of interest.
+ *
+ * @param row - The row index of the cell of interest.
+ *
+ * @param column - The column index of the cell of interest.
+ *
+ * @param returns - The data value for the specified cell.
+ */
+data(region: DataModel.CellRegion, row: number, column: number): any {
+  // Set up the field and value variables.
+  let value: string;
+
+  // Look up the field and value for the region.
+  switch (region) {
+  case 'body':
+    if (this._parsedRowNum !== row) {
+      this._parsedRow = this._data.slice(this._offsets[row], this._offsets[row + 1]).split(this._delimiter);
```

Comment:
Rather than keeping just row offsets, you can keep starting index and length of each datum, then slice the source string each time you need the data. That would eliminate the temporary row array.

Reply:
Are you saying to store the offset/length of *each* datum in the entire data? That means my current 30MB row offset array (in the 8-million-row case I tested) balloons to 30MB*(num of columns)*2.

Reply:
I agree that this is cache is assuming that data is typically asked for in row-major order - not sure how true that is. It did subjectively feel fast in the one test I tried.

Reply:
> I agree that this cache is assuming that data is typically asked for in row-major order - not sure how true that is.

It isn't at all. The current renderer loops over columns, then rows, on the idea that the data is often columnar.